src/features/TakeNow/ui/ValidateTakeNowParams.tsx

```diff
@@ -0,0 +1,180 @@
+          allowDuplicate: false,
+        });
+        errorNotificationRef.current += 1;
+      }
+    },
+    [showError, errorNotificationRef],
+  );
+
+  if (respondentId !== user.id) {
+    // No impersonation allowed... yet?
+    showErrorNotification(t('takeNow.invalidRespondent'));
+    return <ActivityList />;
+  }
+
+  if (isLoadingApplet) {
+    return <Loader />;
+  }
+
+  if (isAppletError || !appletData?.data?.result) {
+    // The applet is part of the URL, so we'll let the activity list page handle this error
+    return <ActivityList />;
+  }
+
+  if (isLoadingSubject || isLoadingActivity) {
+    return <Loader />;
+  }
+
+  if (
+    isSubjectError ||
+    !subjectData?.data?.result ||
+    subjectData.data.result.appletId !== appletId
+  ) {
+    showErrorNotification(t('takeNow.invalidSubject'));
+    return <ActivityList />;
+  }
+
+  if (isActivityError) {
+    showErrorNotification(t('takeNow.invalidActivity'));
+    return <ActivityList />;
+  }
+
+  // At this point we have the subject, applet, and activity data
+  // We can't fetch the workspace roles before we have the applet data
+  // because that's where we get the workspace ID
+
+  if (!workspaceId) {
+    const localWorkspaceId = appletData.data.result.encryption?.accountId;
+    if (!localWorkspaceId) {
+      // We can't verify the roles of the current user in this workspace,
+      // so we can't proceed with a multi-informant flow
+      showErrorNotification(t('common_loading_error'));
+      return <ActivityList />;
+    }
+
+    setWorkspaceId(localWorkspaceId);
+    return <Loader />;
+  }
+
+  if (isLoadingWorkspaceRoles || isLoadingRespondent) {
+    return <Loader />;
+  }
+
+  if (isWorkspaceRolesError || !workspaceRolesData?.data?.result) {
+    // Same as above - we can't verify the roles of the current user in this workspace,
+    // so we can't proceed with a multi-informant flow
+    showErrorNotification(t('common_loading_error'));
+    return <ActivityList />;
+  }
+
+  const roles = workspaceRolesData.data.result[appletId];
+  const hasCorrectRole = TAKE_NOW_ROLES.some((role) => roles.includes(role));
+
+  if (!hasCorrectRole) {
+    showErrorNotification(t('takeNow.invalidRespondent'));
+    return <ActivityList />;
+  }
+
+  if (isRespondentError || !respondentData || !respondentData?.data?.result) {
+    // If we're unable to fetch the subject ID for the current user, we can't start the multi-informant flow
+    // eslint-disable-next-line no-console
+    console.error('Unable to fetch subject ID for current user');
+    return <ActivityList />;
+  }
+
+  const { subjectId: sourceSubjectId } = respondentData.data.result;
+  const multiInformantState = getMultiInformantState();
+
+  if (
+    !multiInformantState ||
+    !multiInformantState.sourceSubjectId ||
+    !multiInformantState.targetSubjectId
+  ) {
+    if (sourceSubjectId !== targetSubjectId) {
+      initiateTakeNow({ sourceSubjectId, targetSubjectId });
```

Comment:
It could lead to an infinite loop without the proper conditions, but since I'm checking for the state before setting it we get to avoid that.

I also don't think it can be totally avoided since hooks have to be called unconditionally and we don't have some of the data when start rendering this component, so we get into a sort of cascade effect and have to do a setState here:

https://github.com/ChildMindInstitute/mindlogger-web-refactor/blob/8d1e0343187ec717b430de04b5c77e0339cbe0a8/src/features/TakeNow/lib/useTakeNowValidation.tsx#L110-L120

We have an upcoming ticket to move this to the backend so we can likely also move this to a side effect then as well, if necessary